docs/adrs/0078-develop-strategy-for-ensuring-code-test-coverage.md

```diff
@@ -0,0 +1,61 @@
+---
+title: 0078 Develop strategy for ensuring code test coverage
+description: |
+  Decision outcome: The Happo Approach
+---
+
+# Develop strategy for ensuring code test coverage
+
+**User Story:** *[MB-16176](https://dp3.atlassian.net/browse/MB-16176)* :lock:
+
+## Problem Statement
+
+Our current system of maintaining test coverage involves measuring the current total test coverage and ensuring that any given PR does not decrease the total coverage.
+This strategy has obvious benefits in helping to support a high test coverage, but it can also introduce unnecessary roadblocks in the PR workflow. To understand this, it's helpful to understand the scenarios that might result in a decrease in total coverage.
+
+
+### Scenarios that result in a test coverage decrease
+1. **Removal of code that is _more_ tested than the total coverage.** If the total coverage is 77.0% and a block of code that is tested at 100% is deleted, the average coverage goes down.
+2. **Addition of code that is _less_ tested than the total coverage and _should be tested more_.** This is the scenario that we want to address with our code coverage strategy.
+3. **Addition of code that is _less_ tested than the total coverage and _doesn't need to be tested more_.** The difference between this scenario and scenario 2 can be subjective, but one can imagine a scenario in which a Go database model is created and doesn't require any tests.
+
+Our current strategy of never allowing the total coverage to decrease attempts to address scenario 2, but it also unfairly flags scenarios 1 and 3. We also have the option to manually reset the test coverage floor, but determining whether that is appropriate (i.e. discerning which scenario is applicable to a given PR) can be difficult and time-consuming. Furthermore, any combination of the three scenarios can be applicable to a given PR -- it doesn't have to be just one -- which can make it extra difficult to decide whether a coverage decrease is legitimate.
```

Comment:
Do we have any data for how often scenario 1 and 3 happen? In my memory, I can recall it happening twice a year which to me doesn't suggest we really have a problem. We have hundreds of PRs a year that the current process ensures we don't have a coverage decrease. Making a change for such an infrequent event doesn't really make a lot of sense to me, personally.

Reply:
I'll also say that I've seen the test coverage problem come up many times, and most of the time the problem is that folks haven't rebased their PR or they need to increase test coverage. IMHO, we should just get rid of the coverage check instead of making it optional, as I believe folks will just ignore it.

Reply:
@ahobson Thanks Drew. Is this supposed to read "**should** just get rid of" _or_ "**shouldn't** just get rid of"?

Reply:
@rogeruiz IMHO, if we think the test coverage problem is major enough to make the checks optional, we should just remove the checks entirely.

Reply:
I've been thinking about this, and I think I'm wrong. I don't have any data to support my position that making the check optional will actually decrease test coverage.

This proposed change is one that we could easily undo if we needed to.

How about this: We note the coverage before putting this change in, and then one month later, at BE/FE check in, we report on the coverage numbers and how they compare to before the change. We can this discuss in the check in if the changes are concerning or not and if we want to keep the checks optional. It can become something we bring up monthly at the check in to ensure we aren't drifting the wrong way.

Thoughts?

Reply:
I updated the PR to record the coverage on merge to main to the govdev environment and we now have a dashboard.

<img width="611" alt="image" src="https://github.com/transcom/mymove-docs/assets/21983/586ccd7f-6a9e-4526-ab28-0c402be26fa4">

Reply:
@ahobson I think this plan makes a lot of sense -- I can mention in the chosen option that we will review it after a month. I should also note that a fairer comparison would be with the coverage if we didn't make the check optional (which we can't know), which would also potentially decrease every time we manually reset it.

Reply:
**thought**: Good point @michaelvaldes-truss. I wonder if to make that comparison match better we could compare the new change in test coverage against historical data of how coverage has changed previously with the required check in place since we have reset the cache a few times.

Reply:
I also want to address the very valid concerns expressed in comments on this PR -- that making the check optional would make it too easy for coverage to decrease. My assessment is that the current strategy already allows for decreases when folks manually reset the limit, and this can happen for very valid reasons (scenarios 2 and 3 described in the ADR). Analyzing coverage decreases is difficult to do automatically, so the intention of the Happo Approach was to put it entirely in the hands of the PR reviewers and also make it easier to access information relevant to that analysis.

I added these points to the ADR and also introduced a third option, Lenience for Tiny Coverage Decreases, which is a middle ground that attempts to partially address these concerns. 

@YanZ777 @kctruss

Reply:
I do want to note that while I support the third option presented - there's a small chance there could be a .2% difference if you pull main to fix a .1% diff in coverage. That's what happened to me this last go round with coverage issues. [Little more detail here.](https://ustcdp3.slack.com/archives/C04U4TWDLKG/p1686263061819939?thread_ts=1686244694.130929&cid=C04U4TWDLKG)